syntax = "proto3";

package grpc;

service ChatApp {
    rpc CreateAccount (CreateAccountRequest) returns (CreateAccountResponse) {}
    rpc Login (LoginRequest) returns (LoginResponse){}
    rpc Logout (LogoutRequest) returns (LogoutResponse){}
    rpc ListAccounts (ListAccountsRequest) returns (ListAccountsResponse) {}
    rpc SendMessage (SendMessageRequest) returns (SendMessageResponse) {}
    rpc DeliverMessages (DeliverRequest) returns (DeliverResponse) {}
    rpc DeleteAccount (DeleteAccountRequest) returns (DeleteAccountResponse) {}
}

message CreateAccountRequest {
    string username = 1;
}

message CreateAccountResponse {
    bool success = 1;
    string errorMessage = 2;
}

message LoginRequest {
    string username = 1;
}

message LoginResponse {
    bool success = 1;
    string errorMessage = 2;
    repeated string queued_messages = 3;
}

message LogoutRequest {
    string username = 1;
}

message LogoutResponse {
    bool success = 1;
    string errorMessage = 2;
}

message ListAccountsRequest {
    string wildcard = 1;
}

message ListAccountsResponse {
    repeated string usernames = 1;
    string errorMessage = 2;
}

message SendMessageRequest {
    string message = 1;
    string froUser = 2;
    string toUser = 3;
}

message SendMessageResponse {
    string message = 1;
    string errorMessage = 2;
}

message DeliverRequest {
    string recipient = 1;
}

message DeliverResponse {
    repeated string messages = 1;
}

message DeleteAccountRequest {
    string username = 1;
}

message DeleteAccountResponse {
    bool success = 1;
    string error = 2;
}